tests/compile/test_fusion_attn.py

```diff
@@ -0,0 +1,159 @@
+        # out = self.attn(q, k, v)
+        # return self.o_proj(out)
+
+        q, k, v = input_.split([self.qkv_size] * 3, dim=-1)
+        return self.attn(q, k, v)
+
+
+class TestModel(torch.nn.Module):
+
+    def __init__(self,
+                 num_layers: int,
+                 head_size=32,
+                 hidden_dim=4096,
+                 num_heads=128,
+                 *args,
+                 **kwargs):
+        super().__init__(*args, **kwargs)
+        self.layers = torch.nn.Sequential(
+            *(TestLayer(head_size, hidden_dim, num_heads)
+              for _ in range(num_layers)))
+
+    def forward(self, input: torch.Tensor):
+        return self.layers(input)
+
+
+PassConfig = CompilationConfig.PassConfig
+
+
+def test_attention_fusion(dist_init):
+    torch.manual_seed(0)
+    torch.set_default_device("cuda")
+
+    input = torch.rand(9, 4096*3)
+
+    enable_attn_fusion = False
+    pass_config = PassConfig(enable_attn_fusion=enable_attn_fusion,
+                             enable_noop=True)
+    compile_config = CompilationConfig(pass_config=pass_config)
+    vllm_config = VllmConfig(compilation_config=compile_config)
+
+    dtype = torch.bfloat16
+    quant_dtype = current_platform.fp8_dtype()
+
+    # head_size = 32
+    # block_size = 16
+    # kvcache_dtype = torch.bfloat16
+    # attn_backend = get_attn_backend(
+    #     head_size,
+    #     dtype,
+    #     None,
+    #     block_size,
+    #     False,
+    # )
+    # impl = attn_backend.get_impl_cls()(0, 0, 0.0, kvcache_dtype)
+    # assert impl.fused_output_quant_supported(quant_dtype,
+    #                                          True, False)
+    # TODO use layerwise impls from layers instead
+
+    # attn_meta_builder = attn_backend.get_builder_cls()()
+
+    attn_metadata = AttentionMetadata(
+        num_prefills=2,
+        num_prefill_tokens=5,
+        num_decode_tokens=4,
+        slot_mapping=torch.arange(0, 9),
+        multi_modal_placeholder_index_maps=None,
+        enable_kv_scales_calculation=False,
+    )
+
+    vllm_config0 = deepcopy(vllm_config)
+    with set_current_vllm_config(vllm_config0), set_forward_context(
+            attn_metadata, vllm_config0, num_tokens=9):
+        # attention layers set the forward context on config
+        model = TestModel(1)
+        out = model(input.clone())
+
+    # set new config so new attn backends (and kv-caches) are created
+    with set_current_vllm_config(vllm_config), set_forward_context(
+            attn_metadata, vllm_config, num_tokens=9):
+        model2 = TestModel(1)
+
+        attn_dict: dict[str, Attention] = vllm_config.compilation_config.static_forward_context
+        assert all(attn_dict[k].impl.fused_output_quant_supported(quant_dtype, True, False) for k in attn_dict)
+        #
+        # passes = []
+        # if enable_attn_fusion:
+        #     passes += [NoOpEliminationPass(vllm_config)] # TODO
+        #     passes += [AttnFusionPass(vllm_config)]
```

Comment:
"TODO: test for AttnFusionPass. I might need help on this as I don't know how to properly set the attention metadata for testing." still a TODO? (Just checking -- I don't know how this works either so I won't be of much help)

Reply:
I got to a running impl but it's giving very wrong results and I don't have bandwidth to debug it right now. I'll get to it when I can.

Reply:
Done!